core/src/main/java/io/substrait/relation/ProtoRelConverter.java

```diff
@@ -0,0 +1,251 @@
+        List<Expression.StructLiteral> structLiterals = new ArrayList<>(virtualTable.getValuesCount());
+        for (io.substrait.proto.Expression.Literal.Struct struct : virtualTable.getValuesList()) {
+            List<Expression.Literal> literals = new ArrayList<>(struct.getFieldsCount());
+            for (io.substrait.proto.Expression.Literal protoLiteral : struct.getFieldsList()) {
+                // FIXME: Ok to expose this or better to create new util?
+                Expression.Literal literal = expressionConverter.from(protoLiteral);
+                literals.add(literal);
+            }
+            ImmutableExpression.StructLiteral structLiteral = ImmutableExpression.StructLiteral.builder().fields(literals).build();
+            structLiterals.add(structLiteral);
+        }
+        return VirtualTableScan.builder().filter(filter).rows(structLiterals).build();
+    }
+
+    private Fetch newFetch(FetchRel fetchRel) {
+        Rel input = from(fetchRel.getInput());
+        return Fetch.builder()
+                .input(input)
+                .count(fetchRel.getCount())
+                .offset(fetchRel.getOffset())
+                .build();
+    }
+
+    private Project newProject(ProjectRel projectRel) {
+        Rel input = from(projectRel.getInput());
+        ProtoExpressionConverter expressionConverter = new ProtoExpressionConverter(lookup, extensions, input.getRecordType());
+        List<Expression> expressions = new ArrayList<>(projectRel.getExpressionsCount());
+        for (io.substrait.proto.Expression protoExpression : projectRel.getExpressionsList()) {
+            Expression expression = expressionConverter.from(protoExpression);
+            expressions.add(expression);
+        }
+        ImmutableProject.Builder builder = Project.builder();
+        return builder.input(input).expressions(expressions).build();
+    }
+
+    private Aggregate newAggregate(AggregateRel aggregateRel) {
+        Rel input = from(aggregateRel.getInput());
+        ProtoExpressionConverter expressionConverter = new ProtoExpressionConverter(lookup, extensions, input.getRecordType());
+        List<Aggregate.Grouping> groupings = new ArrayList<>(aggregateRel.getGroupingsCount());
+        for (AggregateRel.Grouping protoGrouping : aggregateRel.getGroupingsList()) {
+            List<Expression> expressions = new ArrayList<>(protoGrouping.getGroupingExpressionsCount());
+            for (io.substrait.proto.Expression protoExpression : protoGrouping.getGroupingExpressionsList()) {
+                Expression expression = expressionConverter.from(protoExpression);
+                expressions.add(expression);
+            }
+            Aggregate.Grouping grouping = Aggregate.Grouping.builder().expressions(expressions).build();
+            groupings.add(grouping);
+        }
+        List<Aggregate.Measure> measures = new ArrayList<>(aggregateRel.getMeasuresCount());
```

Comment:
For most of these conversions, I'd be inclined to use streams instead of loops. I think it would be easier to read (and probably less code). Any reason you didn't?

For example
```
            List<Expression> expressions = new ArrayList<>(protoGrouping.getGroupingExpressionsCount());
            for (io.substrait.proto.Expression protoExpression : protoGrouping.getGroupingExpressionsList()) {
                Expression expression = expressionConverter.from(protoExpression);
                expressions.add(expression);
            }
```

becomes (roughly)

```
var expressions = protoGrouping.getGroupingExpressionsList().stream().map(expressionConverter::from).toList()
```

Reply:
I've used streams for this method in my last commit. I find it harder to read and debug (partly because I'm not as used to writing the code that way), and in the end same to have the same number of lines. Take a look and let me know what you think. I'm fine either way - I can get used to those APIs too.

Reply:
If I only use streams for the inner loop, it still reads well. I'll push out a version of that.